Mathlib/Tactic/Relation/Trans.lean

```diff
@@ -100,77 +90,68 @@ def getExplicitRelArgCore (tgt rel x z : Expr) : MetaM (Expr × Expr) := do
       getExplicitRelArgCore tgt rel' x z
   | _ => return (rel ,x)
 
+/--
+If `e` of the form `x ~ z`? If so return the triple `(~, x, z)`.
+-/
+def _root_.Lean.Expr.rel? (e : Expr) : MetaM (Option (Expr × Expr × Expr)) := do
+  match e with
+  | Expr.app f z =>
+    match (← getExplicitRelArg? e f z) with
+    | some (rel, x) =>
+      let (rel, x) ← getExplicitRelArgCore e rel x z
+      pure (rel, x, z)
+    | none => pure none
+  | _ => pure none
+
+/--
+Implementation of `Lean.MVarId.trans` and the user-facing `trans` tactic.
+-/
+def _root_.Lean.MVarId.transCore (g : MVarId) (lem : Name) (rel x z : Expr) (y ty : Option Expr) :
+    MetaM (List MVarId) := do
+  let arity ← withReducible <| forallTelescopeReducing (← inferType (← mkConstWithLevelParams lem))
+    fun es _ => pure es.size
+  let y ← y.getDM (mkFreshExprMVar ty)
+  let g₁ ← mkFreshExprMVar (some <| ← mkAppM' rel #[x, y]) .synthetic
+  let g₂ ← mkFreshExprMVar (some <| ← mkAppM' rel #[y, z]) .synthetic
+  g.assign (← mkAppOptM lem (mkArray (arity - 2) none ++ #[some g₁, some g₂]))
+  pure <| [g₁.mvarId!, g₂.mvarId!] ++ (← getMVars y).toList
+
+/--
+Apply a transitivity lemma to the goal, optionally specifying the value `y` of the intermediate
+term, or specifying its type `ty`. (If specifying both the type is not used.)
+-/
+def _root_.Lean.MVarId.trans (g : MVarId) (y ty : Option Expr := none) : MetaM (List MVarId) := do
+  let tgt ← g.getType
+  match ← tgt.rel? with
+  | none => throwError "transitivity lemmas only apply to binary relations, not {indentExpr tgt}"
+  | some (rel, x, z) =>
+    (``Trans.simple :: ``HEq.trans :: (← (transExt.getState (← getEnv)).getUnify rel).toList)
+      |>.firstM fun lem => do g.transCore lem rel x z y ty
+
 /--
 `trans` applies to a goal whose target has the form `t ~ u` where `~` is a transitive relation,
-that is, a relation which has a transitivity lemma tagged with the attribute [trans].
+that is, a relation which has a transitivity lemma tagged with the attribute `@[trans]`.
```

Comment:
Why do we still have both `@[trans]` (used by this tactic) and `class Trans` (used by calc)?

Reply:
Good question, but not relevant here, I think.

Reply:
`class Trans` is picked up by this tactic, so in principle one can only create classes. If the tag `@[trans]` creates an instance of the class (in addition to what it does now), in principle one can migrate smoothly to everything using the class.